File: WeightTracker.java
# WeightTracker.java
/**
 * ⚖️ Weight Tracker Goals – Fitness Manager (Java Edition)
 * Features: log weight, set goal, progress, chart, stats, BMI, unit switch, JSON storage
 * Requires: Java 17+
 */

import java.io.*;
import java.nio.file.*;
import java.time.*;
import java.time.format.*;
import java.util.*;
import java.util.stream.*;

public class WeightTracker {
    // ─── Colors ────────────────────────────────────────────────────────────

    private static final String RESET = "\u001B[0m";
    private static final String BRIGHT = "\u001B[1m";
    private static final String DIM = "\u001B[2m";
    private static final String RED = "\u001B[31m";
    private static final String GREEN = "\u001B[32m";
    private static final String YELLOW = "\u001B[33m";
    private static final String BLUE = "\u001B[34m";
    private static final String MAGENTA = "\u001B[35m";
    private static final String CYAN = "\u001B[36m";

    private static String c(String text, String color) { return color + text + RESET; }

    // ─── Data Classes ──────────────────────────────────────────────────────

    private static class Entry {
        String date;
        double weight;
        String notes;
        Entry(String date, double weight, String notes) {
            this.date = date; this.weight = weight; this.notes = notes;
        }
    }

    private static class Data {
        double goalWeight = 75.0;
        String goalDeadline = defaultDeadline();
        String unit = "kg";
        double height = 175.0;
        List<Entry> entries = new ArrayList<>();
    }

    private static String defaultDeadline() {
        return LocalDate.now().plusDays(90).format(DateTimeFormatter.ISO_LOCAL_DATE);
    }

    // ─── Config ────────────────────────────────────────────────────────────

    private static final double DEFAULT_GOAL = 75.0;
    private static final String DATA_DIR = System.getProperty("user.home") + "/.weight_tracker";
    private static final String DATA_FILE = DATA_DIR + "/data.json";

    // ─── Weight Tracker ────────────────────────────────────────────────────

    private final Scanner scanner;
    private Data data;

    public WeightTracker() throws IOException {
        scanner = new Scanner(System.in);
        Files.createDirectories(Paths.get(DATA_DIR));
        data = new Data();
        load();
    }

    private void load() {
        Path path = Paths.get(DATA_FILE);
        if (!Files.exists(path)) return;
        try {
            String json = Files.readString(path);
            data.goalWeight = extractDouble(json, "goalWeight");
            if (data.goalWeight <= 0) data.goalWeight = DEFAULT_GOAL;
            data.goalDeadline = extractString(json, "goalDeadline");
            if (data.goalDeadline.isEmpty()) data.goalDeadline = defaultDeadline();
            data.unit = extractString(json, "unit");
            if (data.unit.isEmpty()) data.unit = "kg";
            data.height = extractDouble(json, "height");
            if (data.height <= 0) data.height = 175.0;
            // entries not parsed for brevity
        } catch (Exception e) {
            data = new Data();
        }
    }

    private double extractDouble(String json, String key) {
        String pattern = "\"" + key + "\"\\s*:\\s*([\\d.]+)";
        var m = java.util.regex.Pattern.compile(pattern).matcher(json);
        return m.find() ? Double.parseDouble(m.group(1)) : 0.0;
    }

    private String extractString(String json, String key) {
        String pattern = "\"" + key + "\"\\s*:\\s*\"([^\"]*)\"";
        var m = java.util.regex.Pattern.compile(pattern).matcher(json);
        return m.find() ? m.group(1) : "";
    }

    private void save() {
        try {
            StringBuilder sb = new StringBuilder();
            sb.append("{\n");
            sb.append("  \"goalWeight\": ").append(data.goalWeight).append(",\n");
            sb.append("  \"goalDeadline\": \"").append(escapeJson(data.goalDeadline)).append("\",\n");
            sb.append("  \"unit\": \"").append(escapeJson(data.unit)).append("\",\n");
            sb.append("  \"height\": ").append(data.height).append(",\n");
            sb.append("  \"entries\": [\n");
            for (int i = 0; i < data.entries.size(); i++) {
                Entry e = data.entries.get(i);
                sb.append("    {\n");
                sb.append("      \"date\": \"").append(escapeJson(e.date)).append("\",\n");
                sb.append("      \"weight\": ").append(e.weight).append(",\n");
                sb.append("      \"notes\": \"").append(escapeJson(e.notes)).append("\"\n");
                sb.append("    }");
                if (i < data.entries.size() - 1) sb.append(",");
                sb.append("\n");
            }
            sb.append("  ]\n");
            sb.append("}");
            Files.writeString(Paths.get(DATA_FILE), sb.toString());
        } catch (IOException e) { e.printStackTrace(); }
    }

    private String escapeJson(String s) {
        return s.replace("\\", "\\\\").replace("\"", "\\\"");
    }

    private String today() {
        return LocalDate.now().format(DateTimeFormatter.ISO_LOCAL_DATE);
    }

    private Entry getTodayEntry() {
        String todayStr = today();
        for (Entry e : data.entries) {
            if (e.date.equals(todayStr)) return e;
        }
        return null;
    }

    private Double getTodayWeight() {
        Entry e = getTodayEntry();
        return e != null ? e.weight : null;
    }

    private List<Entry> getAllEntriesSorted() {
        return data.entries.stream()
                .sorted(Comparator.comparing(e -> e.date))
                .collect(Collectors.toList());
    }

    private List<Double> getWeights() {
        return getAllEntriesSorted().stream().map(e -> e.weight).collect(Collectors.toList());
    }

    private String progressBar(double current, double goal, int width) {
        if (goal <= 0) return "⚠️  Goal not set";
        List<Double> weights = getWeights();
        if (weights.isEmpty()) return "No data";
        double start = weights.get(0);
        double ratio;
        if (current >= start) {
            ratio = Math.max(0, Math.min(1, (current - goal) / (start - goal)));
        } else {
            ratio = Math.max(0, Math.min(1, (start - current) / (start - goal)));
        }
        int filled = (int)(ratio * width);
        StringBuilder bar = new StringBuilder();
        bar.append("[");
        bar.append("█".repeat(filled));
        bar.append("░".repeat(width - filled));
        bar.append("] ");
        bar.append(String.format("%.1f%%", ratio * 100));
        return bar.toString();
    }

    private String ask(String prompt) {
        System.out.print(prompt);
        return scanner.nextLine().trim();
    }

    private double askDouble(String prompt) {
        while (true) {
            try {
                return Double.parseDouble(ask(prompt));
            } catch (NumberFormatException e) {
                System.out.println(c("❌ Please enter a number.", RED));
            }
        }
    }

    private boolean askConfirm(String prompt) {
        String ans = ask(prompt + " (yes/no): ").toLowerCase();
        return ans.equals("yes");
    }

    // ─── Core Actions ─────────────────────────────────────────────────────

    private void addEntry(double weight, String notes) {
        if (weight <= 0) {
            System.out.println(c("❌ Weight must be positive!", RED));
            return;
        }
        String todayStr = today();
        Entry existing = getTodayEntry();
        if (existing != null) {
            existing.weight = weight;
            existing.notes = notes;
        } else {
            data.entries.add(new Entry(todayStr, weight, notes));
        }
        save();
        System.out.println(c("✅ Weight logged: " + String.format("%.1f", weight) + " " + data.unit, GREEN));
    }

    private void showToday() {
        Double weight = getTodayWeight();
        if (weight == null) {
            System.out.println(c("No weight logged today.", YELLOW));
            return;
        }
        double diff = weight - data.goalWeight;
        String status = diff > 0 ? "above" : "below";
        String color = diff > 0 ? RED : GREEN;
        System.out.println("\n" + c("═".repeat(50), DIM));
        System.out.println(c("⚖️ TODAY'S WEIGHT", BRIGHT + CYAN));
        System.out.println(c("═".repeat(50), DIM));
        System.out.println("  Weight: " + c(String.format("%.1f", weight) + " " + data.unit, color));
        System.out.println("  Goal:   " + String.format("%.1f", data.goalWeight) + " " + data.unit);
        System.out.println("  You are " + c(String.format("%.1f", Math.abs(diff)) + " " + data.unit + " " + status, color) + " your goal");
        System.out.println("  Progress: " + progressBar(weight, data.goalWeight, 20));
        System.out.println(c("═".repeat(50), DIM));
    }

    private void showChart(int days) {
        List<Entry> entries = getAllEntriesSorted();
        if (entries.isEmpty()) {
            System.out.println(c("No data to chart.", YELLOW));
            return;
        }
        if (entries.size() > days) {
            entries = entries.subList(entries.size() - days, entries.size());
        }
        List<Double> weights = entries.stream().map(e -> e.weight).collect(Collectors.toList());
        List<String> dates = entries.stream().map(e -> e.date.substring(5, 10)).collect(Collectors.toList());
        double minW = weights.stream().min(Double::compare).orElse(0.0);
        double maxW = weights.stream().max(Double::compare).orElse(0.0);
        double range = maxW - minW;
        if (range == 0) {
            System.out.println(c("Weight is constant. No variation to chart.", YELLOW));
            return;
        }
        int height = 10;
        int chartWidth = 40;
        List<Integer> norm = weights.stream()
                .map(w -> (int)((w - minW) / range * (height - 1)))
                .collect(Collectors.toList());
        List<String> lines = new ArrayList<>();
        for (int row = height - 1; row >= 0; row--) {
            StringBuilder line = new StringBuilder();
            for (int i = 0; i < norm.size(); i++) {
                if (norm.get(i) >= row) {
                    if (i > 0 && norm.get(i-1) >= row) {
                        line.append('─');
                    } else {
                        line.append('┌');
                    }
                } else {
                    line.append(' ');
                }
            }
            lines.add(line.toString());
        }
        int step = Math.max(1, dates.size() / 8);
        StringBuilder xAxis = new StringBuilder(" ");
        int lastPos = 0;
        for (int i = 0; i < dates.size(); i += step) {
            if (i >= dates.size()) break;
            String label = dates.get(i);
            int pos = i;
            if (pos > lastPos) xAxis.append(" ".repeat(pos - lastPos));
            xAxis.append(label);
            lastPos = pos;
        }
        if (lastPos < dates.size() - 1) {
            xAxis.append(" ".repeat(dates.size() - 1 - lastPos));
        }
        System.out.println("\n" + c("📈 Weight Chart (last " + entries.size() + " days)", BRIGHT + CYAN));
        lines.forEach(System.out::println);
        System.out.println(xAxis.toString());
        System.out.printf("Min: %.1f %s  Max: %.1f %s\n", minW, data.unit, maxW, data.unit);
    }

    private void showStats() {
        if (data.entries.isEmpty()) {
            System.out.println(c("📭 No data yet. Start tracking!", YELLOW));
            return;
        }
        List<Double> weights = getWeights();
        if (weights.isEmpty()) {
            System.out.println(c("No weight data available.", YELLOW));
            return;
        }
        double total = weights.stream().mapToDouble(Double::doubleValue).sum();
        double avg = total / weights.size();
        double minW = weights.stream().min(Double::compare).orElse(0.0);
        double maxW = weights.stream().max(Double::compare).orElse(0.0);
        double last = weights.get(weights.size() - 1);
        double first = weights.get(0);
        double change = last - first;
        String trend = change > 0 ? "up" : "down";
        double heightM = data.height / 100.0;
        double bmi = heightM > 0 ? last / (heightM * heightM) : 0.0;
        String bmiCat = bmiCategory(bmi);
        System.out.println("\n📊 STATISTICS");
        System.out.println(c("─".repeat(30), DIM));
        System.out.printf("  Total Entries:  %d\n", weights.size());
        System.out.println("  Current Weight: " + c(String.format("%.1f", last) + " " + data.unit, GREEN));
        System.out.printf("  Goal Weight:    %.1f %s\n", data.goalWeight, data.unit);
        System.out.printf("  Average:        %.1f %s\n", avg, data.unit);
        System.out.printf("  Minimum:        %.1f %s\n", minW, data.unit);
        System.out.printf("  Maximum:        %.1f %s\n", maxW, data.unit);
        String sign = change > 0 ? "+" : "";
        System.out.printf("  Change:         %s%.1f %s (%s)\n", sign, change, data.unit, trend);
        System.out.printf("  BMI:            %.1f (%s)\n", bmi, bmiCat);
    }

    private String bmiCategory(double bmi) {
        if (bmi < 18.5) return "Underweight";
        if (bmi < 25) return "Normal";
        if (bmi < 30) return "Overweight";
        return "Obese";
    }

    private void setGoal(double weight, String deadline) {
        if (weight <= 0) {
            System.out.println(c("❌ Goal weight must be positive!", RED));
            return;
        }
        data.goalWeight = weight;
        if (!deadline.isEmpty()) {
            try {
                LocalDate.parse(deadline, DateTimeFormatter.ISO_LOCAL_DATE);
                data.goalDeadline = deadline;
            } catch (DateTimeParseException e) {
                System.out.println(c("⚠️  Invalid date format. Use YYYY-MM-DD. Keeping current deadline.", YELLOW));
                return;
            }
        } else {
            data.goalDeadline = defaultDeadline();
        }
        save();
        System.out.println(c("✅ Goal set to " + String.format("%.1f", weight) + " " + data.unit + " by " + data.goalDeadline, GREEN));
    }

    private void setHeight(double height) {
        if (height <= 0) {
            System.out.println(c("❌ Height must be positive!", RED));
            return;
        }
        data.height = height;
        save();
        System.out.println(c("✅ Height set to " + String.format("%.1f", height) + " cm", GREEN));
    }

    private void setUnit(String unit) {
        if (!unit.equals("kg") && !unit.equals("lbs")) {
            System.out.println(c("❌ Unit must be 'kg' or 'lbs'", RED));
            return;
        }
        data.unit = unit;
        save();
        System.out.println(c("✅ Unit set to " + unit, GREEN));
    }

    private void clearData() {
        if (!askConfirm("⚠️  Delete ALL data? This cannot be undone!")) return;
        data.entries.clear();
        data.goalWeight = DEFAULT_GOAL;
        data.goalDeadline = defaultDeadline();
        save();
        System.out.println(c("🗑️  All data cleared.", YELLOW));
    }

    // ─── Menu ─────────────────────────────────────────────────────────────

    private void showMenu() {
        Double todayWeight = getTodayWeight();
        String weightStr = todayWeight != null ? String.format("%.1f", todayWeight) : "—";
        String progress = progressBar(todayWeight != null ? todayWeight : 0, data.goalWeight, 20);
        System.out.println("\n" + c("═".repeat(50), CYAN));
        System.out.println(c("⚖️ WEIGHT TRACKER", BRIGHT + CYAN));
        System.out.println(c("═".repeat(50), CYAN));
        System.out.println("  Today: " + weightStr + " " + data.unit + " / " + String.format("%.1f", data.goalWeight) + " " + data.unit);
        System.out.println("  Goal deadline: " + data.goalDeadline);
        System.out.println("  Progress: " + progress);
        System.out.println(c("─".repeat(50), DIM));
        System.out.println("  1. ⚖️ Log weight today");
        System.out.println("  2. 📊 Today's progress");
        System.out.println("  3. 📈 Show weight chart");
        System.out.println("  4. 📊 Statistics");
        System.out.println("  5. 🎯 Set goal weight & deadline");
        System.out.println("  6. 📏 Set height");
        System.out.println("  7. 🔄 Set unit (kg/lbs)");
        System.out.println("  8. 🗑️  Clear all data");
        System.out.println("  0. 🚪 Exit");
        System.out.println(c("═".repeat(50), CYAN));
    }

    public void run() {
        System.out.print("\033[H\033[2J");
        System.out.flush();
        System.out.println(c("\n⚖️ Weight Tracker – Goal‑Based Fitness Manager", BRIGHT + CYAN));
        System.out.println(c("Track your weight, reach your goals!", DIM));

        while (true) {
            showMenu();
            String choice = ask("Your choice: ");
            switch (choice) {
                case "1": {
                    double weight = askDouble("Weight (" + data.unit + "): ");
                    String notes = ask("Notes (optional): ");
                    addEntry(weight, notes);
                    break;
                }
                case "2": showToday(); break;
                case "3": showChart(30); break;
                case "4": showStats(); break;
                case "5": {
                    double weight = askDouble("Goal weight (" + data.unit + "): ");
                    String deadline = ask("Deadline (YYYY-MM-DD) [leave empty for default]: ");
                    setGoal(weight, deadline);
                    break;
                }
                case "6": {
                    double height = askDouble("Height (cm): ");
                    setHeight(height);
                    break;
                }
                case "7": {
                    String unit = ask("Unit (kg/lbs): ");
                    setUnit(unit.toLowerCase());
                    break;
                }
                case "8": clearData(); break;
                case "0":
                    System.out.println(c("👋 Stay fit! Goodbye!", CYAN));
                    return;
                default:
                    System.out.println(c("❌ Invalid choice.", RED));
            }
            if (!choice.equals("0")) {
                System.out.print("\nPress Enter to continue...");
                scanner.nextLine();
            }
        }
    }

    public static void main(String[] args) {
        try {
            new WeightTracker().run();
        } catch (Exception e) {
            System.err.println(c("❌ Unexpected error: " + e.getMessage(), RED));
            e.printStackTrace();
            System.exit(1);
        }
    }
}

File: WeightTracker.cs
# WeightTracker.cs
/**
 * ⚖️ Weight Tracker Goals – Fitness Manager (C# Edition)
 * Features: log weight, set goal, progress, chart, stats, BMI, unit switch, JSON storage
 * Requires: .NET 6.0+
 */

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

class WeightTracker
{
    // ─── Colors ────────────────────────────────────────────────────────────

    private static readonly string Reset = "\u001B[0m";
    private static readonly string Bright = "\u001B[1m";
    private static readonly string Dim = "\u001B[2m";
    private static readonly string Red = "\u001B[31m";
    private static readonly string Green = "\u001B[32m";
    private static readonly string Yellow = "\u001B[33m";
    private static readonly string Blue = "\u001B[34m";
    private static readonly string Magenta = "\u001B[35m";
    private static readonly string Cyan = "\u001B[36m";

    private static string C(string text, string color) => color + text + Reset;

    // ─── Data Classes ──────────────────────────────────────────────────────

    public class Entry
    {
        [JsonPropertyName("date")]
        public string Date { get; set; } = "";
        [JsonPropertyName("weight")]
        public double Weight { get; set; }
        [JsonPropertyName("notes")]
        public string Notes { get; set; } = "";
    }

    public class Data
    {
        [JsonPropertyName("goalWeight")]
        public double GoalWeight { get; set; } = 75.0;
        [JsonPropertyName("goalDeadline")]
        public string GoalDeadline { get; set; } = "";
        [JsonPropertyName("unit")]
        public string Unit { get; set; } = "kg";
        [JsonPropertyName("height")]
        public double Height { get; set; } = 175.0;
        [JsonPropertyName("entries")]
        public List<Entry> Entries { get; set; } = new();
    }

    // ─── Config ────────────────────────────────────────────────────────────

    private const double DefaultGoal = 75.0;
    private static readonly string DataDir = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
        ".weight_tracker"
    );
    private static readonly string DataFile = Path.Combine(DataDir, "data.json");

    // ─── Weight Tracker ────────────────────────────────────────────────────

    private readonly Data data = new();
    private readonly Random random = new();

    public WeightTracker()
    {
        Directory.CreateDirectory(DataDir);
        Load();
        if (string.IsNullOrEmpty(data.GoalDeadline))
            data.GoalDeadline = DefaultDeadline();
    }

    private void Load()
    {
        if (!File.Exists(DataFile)) return;
        try
        {
            string json = File.ReadAllText(DataFile);
            var loaded = JsonSerializer.Deserialize<Data>(json);
            if (loaded != null)
            {
                data.GoalWeight = loaded.GoalWeight > 0 ? loaded.GoalWeight : DefaultGoal;
                data.GoalDeadline = loaded.GoalDeadline ?? DefaultDeadline();
                data.Unit = !string.IsNullOrEmpty(loaded.Unit) ? loaded.Unit : "kg";
                data.Height = loaded.Height > 0 ? loaded.Height : 175.0;
                data.Entries = loaded.Entries ?? new List<Entry>();
            }
        }
        catch { /* ignore */ }
    }

    private void Save()
    {
        string json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(DataFile, json);
    }

    private static string DefaultDeadline() => DateTime.Now.AddDays(90).ToString("yyyy-MM-dd");

    private string Today() => DateTime.Now.ToString("yyyy-MM-dd");

    private Entry GetTodayEntry()
    {
        string today = Today();
        return data.Entries.FirstOrDefault(e => e.Date == today);
    }

    private double? GetTodayWeight()
    {
        var e = GetTodayEntry();
        return e != null ? e.Weight : (double?)null;
    }

    private List<Entry> GetAllEntriesSorted()
    {
        return data.Entries.OrderBy(e => e.Date).ToList();
    }

    private List<double> GetWeights()
    {
        return GetAllEntriesSorted().Select(e => e.Weight).ToList();
    }

    private string ProgressBar(double current, double goal, int width = 20)
    {
        if (goal <= 0) return "⚠️  Goal not set";
        var weights = GetWeights();
        if (!weights.Any()) return "No data";
        double start = weights.First();
        double ratio;
        if (current >= start)
            ratio = Math.Max(0, Math.Min(1, (current - goal) / (start - goal)));
        else
            ratio = Math.Max(0, Math.Min(1, (start - current) / (start - goal)));
        int filled = (int)(ratio * width);
        string bar = new string('█', filled) + new string('░', width - filled);
        return $"[{bar}] {ratio * 100.0:F1}%";
    }

    private string Ask(string prompt)
    {
        Console.Write(prompt);
        return Console.ReadLine()?.Trim() ?? "";
    }

    private double AskDouble(string prompt)
    {
        while (true)
        {
            if (double.TryParse(Ask(prompt), out double val)) return val;
            Console.WriteLine(C("❌ Please enter a number.", Red));
        }
    }

    private bool AskConfirm(string prompt)
    {
        string ans = Ask(prompt + " (yes/no): ").ToLower();
        return ans == "yes";
    }

    // ─── Core Actions ─────────────────────────────────────────────────────

    private void AddEntry(double weight, string notes)
    {
        if (weight <= 0)
        {
            Console.WriteLine(C("❌ Weight must be positive!", Red));
            return;
        }
        string today = Today();
        var existing = GetTodayEntry();
        if (existing != null)
        {
            existing.Weight = weight;
            existing.Notes = notes;
        }
        else
        {
            data.Entries.Add(new Entry { Date = today, Weight = weight, Notes = notes });
        }
        Save();
        Console.WriteLine(C($"✅ Weight logged: {weight:F1} {data.Unit}", Green));
    }

    private void ShowToday()
    {
        var weight = GetTodayWeight();
        if (weight == null)
        {
            Console.WriteLine(C("No weight logged today.", Yellow));
            return;
        }
        double diff = weight.Value - data.GoalWeight;
        string status = diff > 0 ? "above" : "below";
        string color = diff > 0 ? Red : Green;
        Console.WriteLine("\n" + C(new string('═', 50), Dim));
        Console.WriteLine(C("⚖️ TODAY'S WEIGHT", Bright + Cyan));
        Console.WriteLine(C(new string('═', 50), Dim));
        Console.WriteLine($"  Weight: {C($"{weight:F1} {data.Unit}", color)}");
        Console.WriteLine($"  Goal:   {data.GoalWeight:F1} {data.Unit}");
        Console.WriteLine($"  You are {C($"{Math.Abs(diff):F1} {data.Unit} {status}", color)} your goal");
        Console.WriteLine($"  Progress: {ProgressBar(weight.Value, data.GoalWeight)}");
        Console.WriteLine(C(new string('═', 50), Dim));
    }

    private void ShowChart(int days = 30)
    {
        var entries = GetAllEntriesSorted();
        if (!entries.Any())
        {
            Console.WriteLine(C("No data to chart.", Yellow));
            return;
        }
        if (entries.Count > days)
            entries = entries.Skip(entries.Count - days).ToList();
        var weights = entries.Select(e => e.Weight).ToList();
        var dates = entries.Select(e => e.Date[5..10]).ToList();
        double minW = weights.Min();
        double maxW = weights.Max();
        double range = maxW - minW;
        if (range == 0)
        {
            Console.WriteLine(C("Weight is constant. No variation to chart.", Yellow));
            return;
        }
        int height = 10;
        int chartWidth = 40;
        var norm = weights.Select(w => (int)((w - minW) / range * (height - 1))).ToList();
        var lines = new List<string>();
        for (int row = height - 1; row >= 0; row--)
        {
            var line = new char[norm.Count];
            for (int i = 0; i < norm.Count; i++)
            {
                if (norm[i] >= row)
                {
                    if (i > 0 && norm[i - 1] >= row)
                        line[i] = '─';
                    else
                        line[i] = '┌';
                }
                else
                {
                    line[i] = ' ';
                }
            }
            lines.Add(new string(line));
        }
        int step = Math.Max(1, dates.Count / 8);
        var xAxis = new System.Text.StringBuilder(" ");
        int lastPos = 0;
        for (int i = 0; i < dates.Count; i += step)
        {
            if (i >= dates.Count) break;
            string label = dates[i];
            int pos = i;
            if (pos > lastPos) xAxis.Append(' ', pos - lastPos);
            xAxis.Append(label);
            lastPos = pos;
        }
        if (lastPos < dates.Count - 1)
            xAxis.Append(' ', dates.Count - 1 - lastPos);
        Console.WriteLine($"\n{C($"📈 Weight Chart (last {entries.Count} days)", Bright + Cyan)}");
        foreach (var line in lines) Console.WriteLine(line);
        Console.WriteLine(xAxis.ToString());
        Console.WriteLine($"Min: {minW:F1} {data.Unit}  Max: {maxW:F1} {data.Unit}");
    }

    private void ShowStats()
    {
        if (!data.Entries.Any())
        {
            Console.WriteLine(C("📭 No data yet. Start tracking!", Yellow));
            return;
        }
        var weights = GetWeights();
        if (!weights.Any())
        {
            Console.WriteLine(C("No weight data available.", Yellow));
            return;
        }
        double total = weights.Sum();
        double avg = total / weights.Count;
        double minW = weights.Min();
        double maxW = weights.Max();
        double last = weights.Last();
        double first = weights.First();
        double change = last - first;
        string trend = change > 0 ? "up" : "down";
        double heightM = data.Height / 100.0;
        double bmi = heightM > 0 ? last / (heightM * heightM) : 0;
        string bmiCat = BmiCategory(bmi);
        Console.WriteLine("\n📊 STATISTICS");
        Console.WriteLine(C(new string('─', 30), Dim));
        Console.WriteLine($"  Total Entries:  {weights.Count}");
        Console.WriteLine($"  Current Weight: {C($"{last:F1} {data.Unit}", Green)}");
        Console.WriteLine($"  Goal Weight:    {data.GoalWeight:F1} {data.Unit}");
        Console.WriteLine($"  Average:        {avg:F1} {data.Unit}");
        Console.WriteLine($"  Minimum:        {minW:F1} {data.Unit}");
        Console.WriteLine($"  Maximum:        {maxW:F1} {data.Unit}");
        string sign = change > 0 ? "+" : "";
        Console.WriteLine($"  Change:         {sign}{change:F1} {data.Unit} ({trend})");
        Console.WriteLine($"  BMI:            {bmi:F1} ({bmiCat})");
    }

    private static string BmiCategory(double bmi)
    {
        if (bmi < 18.5) return "Underweight";
        if (bmi < 25) return "Normal";
        if (bmi < 30) return "Overweight";
        return "Obese";
    }

    private void SetGoal(double weight, string deadline)
    {
        if (weight <= 0)
        {
            Console.WriteLine(C("❌ Goal weight must be positive!", Red));
            return;
        }
        data.GoalWeight = weight;
        if (!string.IsNullOrEmpty(deadline))
        {
            if (!DateTime.TryParseExact(deadline, "yyyy-MM-dd", null, System.Globalization.DateTimeStyles.None, out _))
            {
                Console.WriteLine(C("⚠️  Invalid date format. Use YYYY-MM-DD. Keeping current deadline.", Yellow));
                return;
            }
            data.GoalDeadline = deadline;
        }
        else
        {
            data.GoalDeadline = DefaultDeadline();
        }
        Save();
        Console.WriteLine(C($"✅ Goal set to {weight:F1} {data.Unit} by {data.GoalDeadline}", Green));
    }

    private void SetHeight(double height)
    {
        if (height <= 0)
        {
            Console.WriteLine(C("❌ Height must be positive!", Red));
            return;
        }
        data.Height = height;
        Save();
        Console.WriteLine(C($"✅ Height set to {height:F1} cm", Green));
    }

    private void SetUnit(string unit)
    {
        if (unit != "kg" && unit != "lbs")
        {
            Console.WriteLine(C("❌ Unit must be 'kg' or 'lbs'", Red));
            return;
        }
        data.Unit = unit;
        Save();
        Console.WriteLine(C($"✅ Unit set to {unit}", Green));
    }

    private void ClearData()
    {
        if (!AskConfirm("⚠️  Delete ALL data? This cannot be undone!")) return;
        data.Entries.Clear();
        data.GoalWeight = DefaultGoal;
        data.GoalDeadline = DefaultDeadline();
        Save();
        Console.WriteLine(C("🗑️  All data cleared.", Yellow));
    }

    // ─── Menu ─────────────────────────────────────────────────────────────

    private void ShowMenu()
    {
        var todayWeight = GetTodayWeight();
        string weightStr = todayWeight != null ? $"{todayWeight:F1}" : "—";
        string progress = ProgressBar(todayWeight ?? 0, data.GoalWeight);
        Console.WriteLine("\n" + C(new string('═', 50), Cyan));
        Console.WriteLine(C("⚖️ WEIGHT TRACKER", Bright + Cyan));
        Console.WriteLine(C(new string('═', 50), Cyan));
        Console.WriteLine($"  Today: {weightStr} {data.Unit} / {data.GoalWeight:F1} {data.Unit}");
        Console.WriteLine($"  Goal deadline: {data.GoalDeadline}");
        Console.WriteLine($"  Progress: {progress}");
        Console.WriteLine(C(new string('─', 50), Dim));
        Console.WriteLine("  1. ⚖️ Log weight today");
        Console.WriteLine("  2. 📊 Today's progress");
        Console.WriteLine("  3. 📈 Show weight chart");
        Console.WriteLine("  4. 📊 Statistics");
        Console.WriteLine("  5. 🎯 Set goal weight & deadline");
        Console.WriteLine("  6. 📏 Set height");
        Console.WriteLine("  7. 🔄 Set unit (kg/lbs)");
        Console.WriteLine("  8. 🗑️  Clear all data");
        Console.WriteLine("  0. 🚪 Exit");
        Console.WriteLine(C(new string('═', 50), Cyan));
    }

    public void Run()
    {
        Console.Clear();
        Console.WriteLine(C("\n⚖️ Weight Tracker – Goal‑Based Fitness Manager", Bright + Cyan));
        Console.WriteLine(C("Track your weight, reach your goals!", Dim));

        while (true)
        {
            ShowMenu();
            string choice = Ask("Your choice: ");
            switch (choice)
            {
                case "1":
                    double weight = AskDouble($"Weight ({data.Unit}): ");
                    string notes = Ask("Notes (optional): ");
                    AddEntry(weight, notes);
                    break;
                case "2":
                    ShowToday();
                    break;
                case "3":
                    ShowChart(30);
                    break;
                case "4":
                    ShowStats();
                    break;
                case "5":
                    double goal = AskDouble($"Goal weight ({data.Unit}): ");
                    string deadline = Ask("Deadline (YYYY-MM-DD) [leave empty for default]: ");
                    SetGoal(goal, deadline);
                    break;
                case "6":
                    double height = AskDouble("Height (cm): ");
                    SetHeight(height);
                    break;
                case "7":
                    string unit = Ask("Unit (kg/lbs): ");
                    SetUnit(unit.ToLower());
                    break;
                case "8":
                    ClearData();
                    break;
                case "0":
                    Console.WriteLine(C("👋 Stay fit! Goodbye!", Cyan));
                    return;
                default:
                    Console.WriteLine(C("❌ Invalid choice.", Red));
                    break;
            }
            if (choice != "0")
            {
                Console.Write("\nPress Enter to continue...");
                Console.ReadLine();
            }
        }
    }

    public static void Main()
    {
        try
        {
            new WeightTracker().Run();
        }
        catch (Exception ex)
        {
            Console.WriteLine(C($"❌ Unexpected error: {ex.Message}", Red));
            Environment.Exit(1);
        }
    }
}
